File: QytetetJava/src/modeloqytetet/Qytetet.java

package modeloqytetet;

import java.util.ArrayList;
import java.util.List;
import GUIQytetet.*;

public class Qytetet {
    private static final Qytetet INSTANCE = new Qytetet();
    public static final int MAX_JUGADORES = 4;
    static int MAX_CARTAS = 10;
    static int MAX_CASILLAS = 20;
    static int PRECIO_LIBERTAD = 200;
    static int SALDO_SALIDA = 1000;
    private Dado dado;
    private Sorpresa cartaActual;
    private ArrayList<Sorpresa> mazo;
    private Tablero tablero;
    private Jugador jugadorActual;
    private ArrayList<Jugador> jugadores = new ArrayList();
    
    private Qytetet() { 
        this.inicializarCartasSorpresa();
        this.inicializarTablero();   
    }
    
    public static Qytetet getInstance() {
        return INSTANCE;
    }
    
    public Tablero getTablero() {
        return tablero;
    }
    
    public ArrayList<Sorpresa> getMazo() {
        return mazo;
    }
            
    public boolean aplicarSorpresa() {
        boolean tienePropietario = false;
        int valor = cartaActual.getValor();
        
        if (cartaActual.getTipo() == TipoSorpresa.PAGARCOBRAR) {
            jugadorActual.modificarSaldo(valor);
            
            if (valor > 0)
                System.out.println("¡Has recibido dinero! --> " + valor);
            else
                System.out.println("Has perdido dinero --> " + valor);
        }
        else if (cartaActual.getTipo() == TipoSorpresa.IRACASILLA) {
            boolean esCarcel = tablero.esCasillaCarcel(valor);
            
            if (esCarcel) {
                this.encarcelarJugador();
                System.out.println("Vas a la cárcel.");
            }
            else {
               Casilla nuevaCasilla = tablero.obtenerCasillaNumero(valor);
               tienePropietario = jugadorActual.actualizarPosicion(nuevaCasilla);
               System.out.println("Vas a la casilla " + valor);
            }
        }
        else if (cartaActual.getTipo() == TipoSorpresa.PORCASAHOTEL) {
            jugadorActual.pagarCobrarPorCasaYHotel(valor);
            
            if (valor > 0) {
                System.out.println("Ganas " + valor + " por cada casa y hotel en tus propiedades");
            }
            else
                System.out.println("Pierdes " + valor + "por cada casa y hotel en tus propiedades");
        }
        else if (cartaActual.getTipo() == TipoSorpresa.PORJUGADOR) {
            for (Jugador jugador : jugadores) {
                if (jugador.getNombre().equals(jugadorActual.getNombre())) {
                    jugador.modificarSaldo(valor);
                    jugadorActual.modificarSaldo(-valor);
                }
            }
            
            if (valor > 0) {
                System.out.println("Ganas " + valor + " por cada jugador");
            }
            else
                System.out.println("Pierdes " + valor + "por cada jugador"); 
        }
        else if (cartaActual.getTipo() == TipoSorpresa.CONVERTIRME) {
           jugadorActual.convertirme(valor);      
        }
        
        if (cartaActual.getTipo() == TipoSorpresa.SALIRCARCEL) {
            jugadorActual.setCartaLibertad(cartaActual);
            System.out.println("Recibes una carta de libertad");
        }
        else
            mazo.add(cartaActual);
        
        return tienePropietario;
    }
    
    public boolean comprarTituloPropiedad() {
        boolean puedoComprar = jugadorActual.comprarTitulo();
        
        return puedoComprar;
    }
    
    public boolean edificarCasa(Casilla casilla) {
        boolean sePuedeEdificar;
        boolean puedoEdificar = false;
        boolean puedoEdificarCasa;
        int costeEdificarCasa = 0;
        
        if (casilla.soyEdificable()) {
            sePuedeEdificar = casilla.sePuedeEdificarCasa(jugadorActual.factorEspeculador);
            
            if (sePuedeEdificar) {
                puedoEdificar = jugadorActual.puedoEdificarCasa(casilla);
                
                if (puedoEdificar) {
                    costeEdificarCasa = casilla.edificarCasa();
                    jugadorActual.modificarSaldo(-costeEdificarCasa);
                }
            }
        }
        
        puedoEdificarCasa = puedoEdificar;
        
        return puedoEdificarCasa;
    }
    
    public boolean edificarHotel(Casilla casilla) {
        boolean sePuedeEdificar;
        boolean puedoEdificar = false;
        boolean puedoEdificarHotel;
        int costeEdificarHotel = 0;
        
        if (casilla.soyEdificable()) {
            sePuedeEdificar = casilla.sePuedeEdificarHotel(jugadorActual.factorEspeculador);
            
            if (sePuedeEdificar) {
                puedoEdificar = jugadorActual.puedoEdificarHotel(casilla);
                
                if (puedoEdificar) {
                    costeEdificarHotel = casilla.edificarHotel();
                    jugadorActual.modificarSaldo(costeEdificarHotel);
                }
            }
        }
        
        puedoEdificarHotel = puedoEdificar;
        
        return puedoEdificarHotel;
    }
    
    public Sorpresa getCartaActual() {
        return cartaActual;
    }
    
    public Jugador getJugadorActual() {
        return jugadorActual;
    }
    
    public ArrayList<Jugador> getJugadores() {
        return jugadores;
    }
    
    public boolean hipotecarPropiedad(Casilla casilla) {
        boolean puedoHipotecarPropiedad;
        boolean sePuedeHipotecar;
        boolean puedoHipotecar = false;
        int cantidadRecibida;
        
        if (casilla.soyEdificable()) {
            sePuedeHipotecar = !casilla.estaHipotecada();
            
            if (sePuedeHipotecar) {
                puedoHipotecar = jugadorActual.puedoHipotecar(casilla);
            
                if (puedoHipotecar) {
                    cantidadRecibida = casilla.hipotecar();
                    jugadorActual.modificarSaldo(cantidadRecibida);
                }
            }
            
        }
        
        puedoHipotecarPropiedad = puedoHipotecar;
        
        return puedoHipotecarPropiedad;
    }
    
    public boolean cancelarHipoteca(Casilla casilla) {
        boolean puedoCancelarHipoteca;
        int cantidadAPagar = 0;
        
        puedoCancelarHipoteca = jugadorActual.puedoPagarHipoteca(casilla);
        
        if (puedoCancelarHipoteca) {
            cantidadAPagar = casilla.cancelarHipoteca();
            jugadorActual.modificarSaldo(-cantidadAPagar);
        }
        
        return puedoCancelarHipoteca;
    }
    
    public void inicializarJuego(ArrayList<String> nombres) {
        this.inicializarJugadores(nombres);
        this.inicializarCartasSorpresa();
        this.inicializarTablero();
        this.salidaJugadores();
        jugadorActual = jugadores.get(0);
        dado = Dado.getInstance();
        tablero = new Tablero();
    }
    
    public boolean intentarSalirCarcel(MetodoSalirCarcel metodo) {
        boolean libre = false;
        boolean tengoSaldo;
        int valorDado = 0;
        
        if (metodo == MetodoSalirCarcel.TIRANDODADO) {
            dado = Dado.getInstance();
            valorDado = dado.nextNumber();
            libre = valorDado > 5;
            System.out.println("Valor del dado: " + valorDado);
        }
        else if (metodo == MetodoSalirCarcel.PAGANDOLIBERTAD) {
            tengoSaldo = jugadorActual.pagarLibertad(200);
            libre = tengoSaldo;
        }
        
        if (libre) {
            jugadorActual.setEncarcelado(false);
            System.out.println("Te vas de la cárcel.");
        }
        else
            System.out.println("Te quedas un rato más en la cárcel.");
        
        return libre;            
    }
    
    public boolean jugar() {
        int valorDado = 0;
        boolean tienePropietario = false;
        Casilla casillaPosicion, nuevaCasilla;
        
        dado = Dado.getInstance();
        valorDado = dado.nextNumber();
        casillaPosicion = jugadorActual.getCasillaActual();
        nuevaCasilla = tablero.obtenerNuevaCasilla(casillaPosicion, valorDado);
        tienePropietario = jugadorActual.actualizarPosicion(nuevaCasilla);
        
        System.out.println("valor del dado: " + valorDado);
        System.out.println("avanza hasta: " + nuevaCasilla.toString());
        
   
        System.out.println("tienePropietario = " + tienePropietario);
        
        
        if (!nuevaCasilla.soyEdificable()) {
            if (nuevaCasilla.getTipo() == TipoCasilla.JUEZ) {
                this.encarcelarJugador();
            }
            else if (nuevaCasilla.getTipo() == TipoCasilla.SORPRESA) {
                cartaActual = mazo.get(0);
                mazo.remove(0);
            }
        }
        
        return tienePropietario;
    }
    
    public List obtenerRanking() {
        List ranking = new ArrayList();
        int capital;
        
        for (Jugador jugador : jugadores) {
            capital = jugador.obtenerCapital();
            ranking.add(jugador.getNombre()); //Por separado está bien??
            ranking.add(capital);
        }
        
        return ranking;
    }
    
    public ArrayList<Casilla> propiedadesHipotecadasJugador(boolean hipotecadas) {
        ArrayList<TituloPropiedad> propiedades;
        ArrayList<Casilla> casillas = new ArrayList();
        propiedades = jugadorActual.obtenerPropiedadesHipotecadas(hipotecadas);
        
        //Obtengo las casillas a partir de las propiedades
        for (TituloPropiedad t : propiedades)
            casillas.add(t.getCasilla());
        
        return casillas;
    }
    
    public Jugador siguienteJugador() {
        Jugador res;
        int indice;
        
        indice = jugadores.indexOf(jugadorActual);
        
        if (indice == jugadores.size() - 1) { //Si estoy en el último jugador
            indice = 0;
            res = jugadores.get(indice);         //Obtener el primer jugador
        }
        else {
            indice += 1;
            res = jugadores.get(indice); //adelanto una posición
        }
        
        jugadorActual = res;
                
        return res;
    }
    
    public boolean venderPropiedad(Casilla casilla) {
        boolean puedoVender = false;
        
        if (casilla.soyEdificable()) {
            puedoVender = jugadorActual.puedoVenderPropiedad(casilla);
            
            if (puedoVender) {
                jugadorActual.venderPropiedad(casilla);
            }
        }
        
        return puedoVender;
    }
    
    private void encarcelarJugador() {
        if (!jugadorActual.tengoCartaLibertad()) {
            Casilla casillaCarcel = tablero.getCarcel();
            jugadorActual.irACarcel(casillaCarcel);
        }
        else {
            Sorpresa carta = jugadorActual.devolverCartaLibertad();
            mazo.add(carta);
        }
    }
    
    private void inicializarCartasSorpresa() {
       this.mazo = new ArrayList();
       mazo.add(new Sorpresa("Te conviertes en un Especulador",3000,TipoSorpresa.CONVERTIRME));
       mazo.add(new Sorpresa("Te hemos pillado con chanclas y calcetines, lo  sentimos, ¡debes ir a la carcel!", 9, TipoSorpresa.IRACASILLA));
       mazo.add(new Sorpresa("Pagas por la sorpresa",-1500,TipoSorpresa.PAGARCOBRAR));
       mazo.add(new Sorpresa("Ganas por la sorpresa",3000,TipoSorpresa.PAGARCOBRAR));
       mazo.add(new Sorpresa("Vas a la casilla porque...",5,TipoSorpresa.IRACASILLA));
       mazo.add(new Sorpresa("Vas a la carcel por corrupto",9,TipoSorpresa.IRACASILLA));
       mazo.add(new Sorpresa("Pagas por tener tanto y ganar tan poco",2000,TipoSorpresa.PORCASAHOTEL));
       mazo.add(new Sorpresa("A pagar impuestos",1000,TipoSorpresa.PORCASAHOTEL));
       mazo.add(new Sorpresa("Te conviertes en un Especulador",5000,TipoSorpresa.CONVERTIRME));
       mazo.add(new Sorpresa("Tus amigos estan generosos",300,TipoSorpresa.PORJUGADOR));
       mazo.add(new Sorpresa("Le debes pasta a tus colegas",-250,TipoSorpresa.PORJUGADOR));
       mazo.add(new Sorpresa("Vas a la casilla porque...",15,TipoSorpresa.IRACASILLA));
       mazo.add(new Sorpresa("Una pena que te vallas, sigue asi y volveras",0,TipoSorpresa.SALIRCARCEL));

    }
    
    private void inicializarJugadores(ArrayList<String> nombres) {
        for (String n: nombres)
            jugadores.add(new Jugador(n));
    }
    
    private void inicializarTablero() {
        tablero = new Tablero();
    }
    
    private void salidaJugadores() {
        for (Jugador j : jugadores)
            j.setCasillaActual(new OtraCasilla(0, TipoCasilla.SALIDA));
    }
    
}

File: QytetetRuby/lib/qytetet.rb
#encoding: utf-8
require "singleton"
require_relative "sorpresa"
require_relative "jugador"
require_relative "tipo_sorpresa"
require_relative "tipo_casilla"
require_relative "tablero"
require_relative "casilla"
require_relative "titulo_propiedad"
require_relative "metodo_salir_carcel"
require_relative "dado"

module ModeloQytetet
  class Qytetet
    include Singleton
    
    attr_accessor :cartaActual, :jugadorActual , :MAX_JUGADORES
    attr_accessor :MAX_CARTAS, :MAX_CASILLAS, :PRECIO_LIBERTAD, :SALDO_SALIDA
    attr_reader :jugadores, :tablero, :mazo
    
    @@MAX_JUGADORES = 4
    @@MAX_CARTAS = 10
    @@MAX_CASILLAS = 20
    @@PRECIO_LIBERTAD = 200
    @@SALDO_SALIDA = 1000
    
    def initialize
#      @jugadores = Array.new
#      @cartaActual = nil
#      @dado = Dado.instance
#      @tablero = Tablero.new
    end
    
    
    def aplicar_sorpresa
      tiene_propietario = false
      valor = @cartaActual.valor
      
      if @cartaActual.tipo == TipoSorpresa::PAGARCOBRAR then
        @jugadorActual.modificar_saldo(valor)
        
        if valor > 0 then
          puts("¡Has recibido dinero! --> " + valor.to_s)
        else
          puts("Has perdido dinero --> #{valor}")
        end
        
      elsif @cartaActual.tipo == TipoSorpresa::IRACASILLA then 
        es_carcel = @tablero.es_casilla_carcel(valor)
        
        if es_carcel then
          encarcelar_jugador
          puts("Vas a la cárcel")
        else
          nueva_casilla = @tablero.obtener_casilla_numero(valor, desplazamiento)
          tiene_propietario = @jugadorActual.actualizar_posicion(nueva_casilla)
          puts("Vas a la casilla #{valor}")
        end
      elsif @cartaActual.tipo == TipoSorpresa::PORCASAHOTEL then
        @jugadorActual.pagar_cobrar_por_casa_y_hotel(valor)
        
        if valor > 0 then
          puts("Ganas #{valor}")
        else
          puts("Pierdes #{valor}")
        end
      elsif @cartaActual.tipo == TipoSorpresa::PORJUGADOR then
        @jugadores.each do |jugador|
          if jugador != @jugadorActual then
            jugador.modificar_saldo(valor)
            @jugadorActual.modificar_saldo(-valor)
          end
        end
      elsif @cartaActual.tipo == TipoSorpresa::PORJUGADOR then
        @jugadorActual.convertirme(valor)
        
        if valor > 0 then
          puts("Ganas #{valor}")
        else
          puts("Pierdes #{valor}")
        end
      end
      
      if @cartaActual.tipo == TipoSorpresa::SALIRCARCEL then
        @jugadorActual.set_carta_libertad(@cartaActual)
        puts("Recibes una carta de libertad")
      else
        @mazo << @cartaActual
      end
      
      tiene_propietario
    end
    
    def cancelar_hipoteca(casilla)
      puedo_cancelar_hipoteca = @jugadorActual.puedo_pagar_hipoteca(casilla)
      
      puedo_cancelar_hipoteca
    end
    
    def comprar_titulo_propiedad
      puedo_comprar = @jugadorActual.comprar_titulo
    
      puedo_comprar
    end
    
    def edificar_casa(casilla)
      puedo_edificar = false
      coste_edificar_casa = 0
      
      if casilla.soy_edificable then
        se_puede_edificar = casilla.se_puede_edificar_casa(@jugadorActual.get_factor_especulador)
        
        if se_puede_edificar then
          puedo_edificar = @jugadorActual.puedo_edificar_casa(casilla)
          
          if puedo_edificar then
            coste_edificar_casa = casilla.edificar_casa
            @jugadorActual.modificar_saldo(-coste_edificar_casa)
          end
        end
      end
      
      puedo_edificar_casa = puedo_edificar
      
      puedo_edificar_casa
    end
    
    def edificar_hotel(casilla)
      puedo_edificar = false
      coste_edificar_hotel = 0
      
      if casilla.soy_edificable then
        se_puede_edificar = casilla.se_puede_edificar_hotel(@jugadorActual.get_factor_especulador)
        
        if se_puede_edificar then
          puedo_edificar = @jugadorActual.puedo_edificar_hotel(casilla)
          
          if puedo_edificar then
            coste_edificar_hotel = casilla.edificar_hotel
            @jugadorActual.modificar_saldo(-coste_edificar_hotel)
          end
        end
      end
      
      puedo_edificar_hotel = puedo_edificar
      
      puedo_edificar_hotel
    end
    
    def get_carta_actual
      @cartaActual
    end
    
    def get_jugador_actual
      @jugadorActual
    end
    
    def get_jugadores
      @jugadores
    end
        
    def hipotecar_propiedad(casilla)
      puedo_hipotecar = false
      
      if casilla.soy_edificable then
        se_puede_hipotecar = !casilla.esta_hipotecada
        
        if se_puede_hipotecar then
          puedo_hipotecar = @jugadorActual.puedo_hipotecar(casilla)
          
          if puedo_hipotecar then
            cantidad_recibida = casilla.hipotecar
            @jugadorActual.modificar_saldo(cantidad_recibida)
          end
        end
      end
      
      puedo_hipotecar_propiedad = puedo_hipotecar
      
      puedo_hipotecar_propiedad
    end
    
    def inicializar_juego(nombres)
      @jugadores = Array.new
      @cartaActual = nil
      @dado = Dado.instance
      @tablero = Tablero.new
      
      inicializar_jugadores(nombres)
      inicializar_cartas_sorpresa
      inicializar_tablero
      salida_jugadores
      @jugadorActual = @jugadores.at(0)
      @cartaActual = @jugadorActual
    end
    
    def intentar_salir_carcel(metodo)
      libre = false
      tengo_saldo = false
      valor_dado = 0
        
      if metodo == MetodoSalirCarcel::TIRANDODADO then
        valor_dado = @dado.tirar
        libre = valor_dado > 5
        puts("Valor del dado: #{valor_dado}")
      elsif metodo == MetodoSalirCarcel::PAGANDOLIBERTAD then
        tengo_saldo = @jugadorActual.pagar_libertad(200)
        libre = tengo_saldo
      end
      
      if libre then
        @jugadorActual.set_encarcelado(false)
        puts("Te vas de la cárcel.")
      else
        puts("Te quedas un rato más en la cárcel.")
      end
      
      libre
    end
    
    def jugar
      valor_dado = 0
      tiene_propietario = false
      
      valor_dado = @dado.tirar
      casilla_posicion = @jugadorActual.casillaActual
      nueva_casilla = @tablero.obtener_nueva_casilla(casilla_posicion, valor_dado)
      tiene_propietario = @jugadorActual.actualizar_posicion(nueva_casilla)
      
      puts("Valor del dado: #{valor_dado}")
      puts("Avanza hasta: #{nueva_casilla.to_s}")
      
      puts ("tiene_propietario = #{tiene_propietario}")
      
      if !nueva_casilla.soy_edificable then
        if nueva_casilla.tipo == TipoCasilla::JUEZ then
          encarcelar_jugador
        elsif nueva_casilla.tipo == TipoCasilla::SORPRESA then
          @cartaActual = @mazo.at(0)
          @mazo.delete(0)
        end
      end
      
      tiene_propietario
    end
    
    def obtener_ranking
      ranking = List.new
      capital = 0
      
      @jugadores.each { |jugador|
        capital = jugador.obtener_capital
        ranking << jugador.nombre
        ranking << capital
      }
      
      ranking
    end
    
    def propiedades_hipotecadas_jugador(hipotecadas)
      propiedades = Array.new
      casillas = Array.new
      propiedades = @jugadorActual.obtener_propiedades_hipotecadas(hipotecadas)
      
      propiedades.each { |t|  
        casillas << t.casilla
      }
      
      casillas
    end
    
    def siguiente_jugador
      indice = @jugadores.index(@jugadorActual)
      
      if indice == (@jugadores.length - 1) then
        indice = 0
      else
        indice = indice + 1
      end
      
      res = @jugadores.at(indice)
      
      @jugadorActual = res
      
      res
    end
    
    def vender_propiedad(casilla)
      puedo_vender = false
      
      if casilla.soy_edificable then
        puedo_vender = @jugadorActual.puedo_vender_propiedad(casilla)
        
        if puedo_vender then
          @jugadorActual.vender_propiedad(casilla)
        end
      end
      
      puedo_vender
    end
    
    def encarcelar_jugador
      if !@jugadorActual.tengo_carta_libertad then
        casilla_carcel = @tablero.get_carcel
        @jugadorActual.ir_a_carcel(casilla_carcel)
      
      else
        carta = @jugadorActual.devolver_carta_libertad
        @mazo << carta
      end
    end
    
    def inicializar_cartas_sorpresa
      @mazo = Array.new
      @mazo = []
      t = Tablero.new
      @mazo<< Sorpresa.new("Te hemos pillado con chanclas y calcetines, lo  sentimos, debes ir a la carcel", 9, TipoSorpresa::IRACASILLA)
      @mazo<< Sorpresa.new("Pagas por la sorpresa",-1500, TipoSorpresa::PAGARCOBRAR)
      @mazo<< Sorpresa.new("Ganas por la sorpresa",3000, TipoSorpresa::PAGARCOBRAR)
      @mazo<< Sorpresa.new("Vas a la casilla porque...",5, TipoSorpresa::IRACASILLA)
      @mazo<< Sorpresa.new("Vas a la carcel por corrupto", t.get_carcel.get_numero_casilla, TipoSorpresa::IRACASILLA)
      @mazo<< Sorpresa.new("Pagas por tener tanto y ganar tan poco",2000, TipoSorpresa::PORCASAHOTEL)
      @mazo<< Sorpresa.new("Te conviertes en Especulador",5000, TipoSorpresa::CONVERTIRME)
      @mazo<< Sorpresa.new("A pagar impuestos",1000, TipoSorpresa::PORCASAHOTEL)
      @mazo<< Sorpresa.new("Tus amigos estan generosos",300, TipoSorpresa::PORJUGADOR)
      @mazo<< Sorpresa.new("Le debes pasta a tus colegas",-250, TipoSorpresa::PORJUGADOR)
      @mazo<< Sorpresa.new("Vas a la casilla porque...",15, TipoSorpresa::IRACASILLA)
      @mazo<< Sorpresa.new("Una pena que te vallas, sigue asi y volveras",0, TipoSorpresa::SALIRCARCEL)
      @mazo<< Sorpresa.new("Te conviertes en Especulador",3000, TipoSorpresa::CONVERTIRME)

      #puts("Inicialicé cartas sorpresa")
    end
        
    def inicializar_jugadores(nombres)
      #puts("Voy a inicializar jugadores")
      @jugadores = []
      
      i = 0
      while i < nombres.length do
        @jugadores << Jugador.new_jugador(nombres.at(i))
        i = i + 1
      end 
      
      #puts("Inicialicé jugadores")
    end
    
    def inicializar_tablero
      @tablero = Tablero.new
      #puts("Inicialicé el tablero")
    end
    
    def salida_jugadores
      @jugadores.each do |j| 
        j.set_casilla_actual(Casilla.new(0, TipoCasilla::SALIDA))
        
      end
      #puts("Jugadores en casilla de salida")
    end
    
    def get_jugadores
      @jugadores
    end
    
    def max_jugadores
      @@MAX_JUGADORES
    end
    
    private :encarcelar_jugador, :inicializar_cartas_sorpresa, :inicializar_jugadores, :inicializar_tablero, :salida_jugadores
  end
end
